### PR TITLE
Raster colorization via `"raster-color"`

### PR DESCRIPTION
**Pull request by rreusser**
This PR implements a `raster-color` style spec property as originally requested in #3889. It is closely analogous in both usage and implementation to [heatmap-color](https://docs.mapbox.com/mapbox-gl-js/style-spec/layers/#paint-heatmap-heatmap-color). The essence of this feature is to reduce an RGBA pixel input to a scalar value, then to map that to a one-dimensional color scale. For efficiency, a lookup table is tabulated on the CPU and then sampled on the GPU, so that performance impact should be minimal.

This PR implements the following style-spec paint properties:

- `raster-color` (`expression`): defines a color map by which a raster layer is colorized. Must implement a style spec expression which operates on `["raster-value"]` and returns a [color](https://docs.mapbox.com/mapbox-gl-js/style-spec/types/#color).
- `raster-color-mix` (`number[4]`): specifies the combination of red, green, and blue channels used to reduce input to a scalar value for color map lookup. The scalar value is computed from the source color `src` using the formula: `src.r * mix.r + src.g * mix.g + src.b * mix.b + mix.a`. Default is luminosity, i.e. `[ 0.2126, 0.7152, 0.0722, 0 ]`.
- `raster-color-range` (`number[2]`): defines the range in which the color map is tabulated. The raster-color expression is evaluated at 1024 steps (non-configurable) uniformly distributed in the range specified by raster-color-range.

The video below cycles through a number of use cases highlighted by the demo page, in particular:
1. Colorization of satellite imagery
2. Colorization of grayscale radar tiles ([rreusser.b3c8wj9i](https://studio.mapbox.com/tilesets/rreusser.b3c8wj9i/), [rreusser.5vzfcdjn](https://studio.mapbox.com/tilesets/rreusser.5vzfcdjn/))
3. Colorization of rgb-encoded terrain tiles ([mapbox.terrain-rgb](https://studio.mapbox.com/tilesets/mapbox.terrain-rgb/))
4. Colorization of rgb-encoded population density tiles ([rreusser.796swou6](https://studio.mapbox.com/tilesets/rreusser.796swou6/))
5. Mapping of grayscale cloud layer ([rreusser.dv2uv1dw](https://studio.mapbox.com/tilesets/rreusser.dv2uv1dw/)) to transparent white layer (works with video layers as well, but omitted here for size; see, for example, [the globe view cloud cover demo](https://www.mapbox.com/blog/globe-view#d3-globe-flex), which raster-colorization significantly simplifies)

*Note: named color scales are not part of this feature. For a demo page, I've tabulated them using d3.*

https://user-images.githubusercontent.com/572717/200716983-2158a873-dd65-439e-ac35-b99a980fa0e2.mov

## Hypsometric shading example

I think an example will make usage most clear. Let's consider what's basically the most nontrivial example, in which data is rgb-encoded. (RGB-encoding is _not_ a requirement, and grayscale inputs permitting 256 levels, for example, require only specification of `raster-color` to style.) Terrain tiles are [RGB-encoded](https://docs.mapbox.com/data/tilesets/reference/mapbox-terrain-rgb-v1/) via [the formula](https://docs.mapbox.com/data/tilesets/reference/mapbox-terrain-rgb-v1/#elevation-data) `height = -10000 + ((R * 256 * 256 + G * 256 + B) * 0.1)`. To decode elevation, we can specify:

```js
'raster-color-mix': [
  256 * 256 * 256 * 0.1,
  256 * 256 * 0.1,
  256 * 0.1,
  -10000
]
```

Note the extra factors of 256. This happens because [0, 255] is normalized to [0, 1] in the shader. We now have an elevation to work with, in meters. We can define the `raster-color` property:

```js
'raster-color': [ "interpolate",
  ["linear"],
  ["raster-value"],
  0, "rgb(57, 143, 83)",
  500, "rgb(178, 205, 174)",
  1000, "rgb(221, 207, 153)",
  2000, "rgb(207, 155, 103)",
  4000, "rgb(227, 210, 197)",
  8848, "rgb(255, 255, 255)"
]
```

Note that the color map is parameterized in terms of _meters_, from sea level (0 meters) to mount everest (8848 meters). Then finally, we direct it to tabulate the color map in this range:

```js
'raster-color-range': [0, 8848]
```

The final note is that RGB-encoded values _must_ be sampled from uncompressed png tiles. For historical reasons, such tiles must be requested from Mapbox using the extension `.pngraw`. Furthermore, Mapbox may overwrite extensions when requesting tiles in order to minimize data transfer, so that you must use `transformRequest` to force requesting of PNG tiles. We must also avoid resampling, so that when requesting tiles from Mapbox we remove `'@2x'` as well. For example:

```
new mapboxgl.Map({
  ...,
  transformRequest: function (url, type) {
    if (type !== "Tile") return;
    return { url: url.replace('.webp', '.pngraw').replace('@2x', '') };
  }
})
```

The result looks something like this:

Screen Shot 2022-11-08 at 6 46 16 PM

For many uses like colorization of grayscale imagery, the default values of `raster-color-mix` and `raster-color-range` would be fine and only `raster-color` would need to be specified.

## Additional notes
- [This comment](https://github.com/mapbox/mapbox-gl-js/issues/3889#issuecomment-521171611) pointed out that it's not pleasant to constantly transform between a normalized color map range (e.g [0, 1]) and physical values. This PR includes `raster-color-range` so that you can define the range over which the color map is defined and evaluated.
- Source alpha is *not* used in the `raster-color-mix` formula. Instead, the source alpha (that is, the alpha value of a tile image pixel) is carried through and applied as opacity to the final, colorized tile. The fourth component of `raster-color-mix` is instead always added as a constant offset. This feels a bit clumsy, but is what enables straightforward decoding of rgb-encoded data. For terrain, it would be the -10000 offset. See, for example, the above terrain example.
- Instead of first reducing to a scalar and colorizing in 1D, fully general 3D mapping of color, (r, g, b) -> (r', g', b'), was considered. However, implementation and usage are more complicated and the 3D color lookup table requires significantly more memory. 
- This feature modifies the raster shader and so works wherever the raster shader is used. It therefore works with projections and video layers.

cc @mapbox/gl-native (It's my intent to work on implementation if/when we feel good about this PR. I believe all shader changes are contained by ifdef's so that this should be native-compatible, with the exception of me needing to figure out how to skip render tests.)
cc @mapbox/map-design-team for initial implementation

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] include before/after visuals or gifs if this PR includes visual changes
 - [x] write tests for all new functionality
 - [ ] document any changes to public APIs
 - [ ] post benchmark scores
 - [x] manually test the debug page
 - [x] tagged `@mapbox/map-design-team` `@mapbox/static-apis` if this PR includes style spec API or visual changes
 - [x] tagged `@mapbox/gl-native` if this PR includes shader changes or needs a native port
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>Add raster colorization for raster layers</changelog>`

### CODE REVIEW

File: debug/raster-color.html
Comment (rreusser) [inline debug/raster-color.html:124]: Included the sake of illustration and debugging, but we should probably remove tilesets on my account from the final PR.

File: src/style-spec/reference/v8.json
Comment (rreusser) [inline src/style-spec/reference/v8.json:6100]: Oops. This is unfinished. These parameter docs need to be thought out and described more completely.

File: src/render/program/raster_program.js
Comment (mourner) [inline src/render/program/raster_program.js:79]: Nit: it would be nice to avoid setting these two to NaNs if there's no colorization, just to make sure there's zero overhead for usual raster layers.

**Review by mourner — COMMENTED** [review]
This looks amazing, and such a cool feature for so little code! Especially love the API design and a big selection of demos. Great work @rreusser, let's get this in!

File: src/render/program/raster_program.js
Comment (rreusser) [reply src/render/program/raster_program.js:79]: Good catch. I think it's the case that uniforms will not be set if they are not specified in the shader. See:

https://github.com/mapbox/mapbox-gl-js/blob/81981ae18d5ea8015203415143083fafd49bb8ee/src/render/uniform_binding.js#L53

To that end, this commit moved the `uniform` definitions _inside_ `#ifdef RASTER_COLOR`, which I think will be sufficient to prevent this uniform from being set when raster color is not active. 
 c4fbb785973a9646221c0f0bd0107ee904efe61f

**Comment by endanke** [comment]
From gl-native perspective it looks good! There isn't too many changes and I can port it right away after it's merged.

File: src/render/draw_raster.js
Comment (endanke) [inline src/render/draw_raster.js:121]: If we only pass 2 for `u_color_ramp` do we really need an uniform for it or could we just hardcode it in the shader? Or is it needed for some future use-cases?

File: src/shaders/raster.fragment.glsl
Comment (endanke) [inline src/shaders/raster.fragment.glsl:36]: Please move the definition of `vec4 color` before the `#ifdef` because we convert these blocks to regular if block in Metal. (Eventually we'll add some checks for the gl-js repo for these formatting requirements.)

So:
```
vec4 color;
#ifdef RASTER_COLOR
    color = ...
```
And the same variable can be used in the else branch.

File: src/style-spec/reference/v8.json
Comment (endanke) [inline src/style-spec/reference/v8.json:6094]: How did you choose these default values?

Comment (endanke) [reply src/style-spec/reference/v8.json:6094]: Ah okay I see it in the PR description. Maybe it could be mentioned in the documentation section of this property?

Comment (rreusser) [reply src/style-spec/reference/v8.json:6094]: Ah yes, for RGB imagery input, luminosity seemed, visually at least, like the least surprising starting point. And note that for grayscale input (e.g. b&w [clouds](https://studio.mapbox.com/tilesets/rreusser.dv2uv1dw/#0.86/0/0) or [radar](https://studio.mapbox.com/tilesets/rreusser.5vzfcdjn/#0.86/0/-97.5)), the components add up to one so that [1, 0, 0, 0] = [0, 1, 0, 0] = [0, 0, 1, 0] = [1/3, 1/3, 1/3, 0] = [0.2126, 0.7152, 0.0722, 0] are all equivalent.

Good point. I'll add a note 👍

**Comment by rreusser** [comment]
@endanke Thanks for the review! 🙇 Nothing is set in stone yet, so if you have any other opinions on the design, nothing is off the table. Points of contention/compromise, as I see it:

- Supports linear combinations only via `raster-color-mix`, so that in-shader log-scaling of data (or similar) is not supported
- The mapping only supports RGB -> scalar -> R'G'B'. @kkaefer suggested maybe a fully general RGB -> R'G'B' mapping using a 256x256x256 lookup table, but we opted for the simpler approach.
- The fourth component of `raster-color-mix` stands out in that it's just an offset which is not multiplied by source alpha. This makes decoding RGB-encoded data easy but feels just a bit low-level. I think it's okay though and just needs clear explanation or presets in studio. 🤷 

Anyway, just wanted to get all the ideas and decisions on the table before etching it in stone. Thanks again!

File: src/render/draw_raster.js
Comment (rreusser) [reply src/render/draw_raster.js:121]: Upon further investigation, you can't hard-code the sampler index inside the shader. It has to be a uniform. I've at least moved this to a js `const` declaration. Note that per previous discussion, WebGL is smart enough not to set uniforms if the uniform is inside an inactive preprocessor declaration, so that it at least shouldn't have meaningful extra cost.

**Review by SnailBones — APPROVED** [review]
Amazing feature @rreusser! 😍 

One question: would it be possible to by default infer `raster-color-range` from the expression used in `raster-color`? Or would this be nontrivial given the way expressions work?

**Comment by rreusser** [comment]
@SnailBones I think it's possible, but the default value of `raster-color-range` would then have to be `null` instead of `[0, 1]` so that we could differentiate between the default [0, 1] as an affirmatively set value and as an unspecified value, which I guess would have to be applied later, where used. I get where you're coming from, but unless there's a good precedent for this, I think it would cause downstream uses like studio to have to know that there are two default values: on in the style spec, and one applied after the fact in software.

**Comment by rreusser** [comment]
One main thing has been holding this up: ensuring this will handle everything we want to throw at it. It's been suggested to use this for _categorical_ data, so I've outlined below why this may be challenging.

## How does categorical data differ from quantitative?

Categorical data means different input values are mapped to different colors, but that the order has no meaning. It's an _encoding_ but not necessarily an _ordering_. Catergories A, B, and C may correspond to RGB values 0, 127, and 255, but they could equally well be reordered. _Quantitative_ data, of course can be mapped to 0-255 but the ordering of each value _defines_ the meaning.

## Why does this matter?

Colorization of categorical data results in artifacts a the boundary between categories.

The reason comes down to _order of operations_. There are two operations relevant here: 1) interpolation, and 2) color scale sampling. The order in which these are performed makes a big difference.

As a test, I made a simple texture with solid colors and no anti-aliasing and applied the [d3 spectral color scale](https://github.com/d3/d3-scale-chromatic#schemeSpectral).





## Desired output

We expect simple colorization, e.g.:

Screenshot 2023-01-03 at 10 47 40 AM


## Actual output

What we actually get is boundaries that traverse the entire color scale between categories, rather than categories meeting smoothly.

Screenshot 2023-01-03 at 10 47 27 AM

Why does this happen? The average of black and white is 50% gray, so when we look up that color scale value, it could be anything. Below, input rgb(0,0,0) meets rgb(255,255,255). Due to linear interpolation, the result traverses the entire color scale and we get a whole rainbow where what we really want is direct interpolation from red to purple.

Screenshot 2023-01-03 at 10 52 46 AM

## Where does this interpolation happen?

This interpolation happens in _two_ places:
1. MTS. During tiling of GeoTIFF or similar input, colors are interpolated linearly. For this reason, it's avoided at a low level for terrain-rgb but can't be turned off for general tile sets.
2. Raster paint properties. Shader interpolation can simply be turned off with [raster-resampling](https://docs.mapbox.com/mapbox-gl-js/style-spec/layers/#paint-raster-raster-resampling): `'raster-resampling': 'nearest'`

So deactivation at the Mapbox GL layer level works fine _if you have full control over the tiling_. If you don't control tiling, then this problematic interpolation is baked into the tileset and there's no going back.

## Can we work with this?

If we'd like the ability to display categorical data, then raster-sampling must be set to `nearest` and we must control the tiling process to avoid baking interpolation into the tileset. If those conditions are met, then it is possible to colorize categorical data.

**Comment by AlexanderBelokon** [comment]
Do we actually want direct interpolation from red to purple? Wouldn't it make more sense to have nearest neighbor or something like [hqx](https://en.wikipedia.org/wiki/Hqx) for categorical data instead?

Also non-linear interpolations in MTS is a good feature to have in any case, since linear interpolation looks nasty, and the price for a nicer interpolation only has to be paid once. Is that on the roadmap at all? If it is, we could add `nearest` to the list of options too.

**Comment by rreusser** [comment]
@AlexanderBelokon That's a good point. The above presumes that nearest neighbor sampling is acceptable. Nicer sampling is, of course, very much preferable, but in order to accomplish that I think we'd have to (re)implement linear or better filtering _in the shader_, after colorization, as opposed to relying on built-in linear texture filtering.

**Comment by rreusser** [comment]
This PR will happen, but we're putting it temporarily on hold as we more carefully consider the API design before permanently committing to it.

File: src/style-spec/reference/v8.json
Comment (pozdnyakov) [inline src/style-spec/reference/v8.json:6074]: @rreusser should we provide a default value similarly to heatmap-color
```
      "default": [
        "interpolate",
        [
          "linear"
        ],
        [
          "heatmap-density"
        ],
        0,
        "rgba(0, 0, 255, 0)",
        0.1,
        "royalblue",
        0.3,
        "cyan",
        0.5,
        "lime",
        0.7,
        "yellow",
        1,
        "red"
      ],
```

Comment (rreusser) [reply src/style-spec/reference/v8.json:6074]: I don't have strong feelings on the default. I thought maybe grayscale was reasonable because we can really just say a lot less about the meaning of the data than we can for heatmap (which is necessarily point density).

**Comment by wboykinm** [comment]
@stepankuzmin A status question for you on this PR: _what do you think remains here before we can merge?_

We've marked this as [one of the components of our new raster arrays pipeline](https://docs.google.com/document/d/1Ta7y_sJHoV-483yDwqXNNimHYk_483MteHtPvROTQNI/edit?usp=sharing), and merging this will be a key step.

cc @jqtrde

**Comment by ghalter** [comment]
Hi, just wanted to ask if there is any news on this or an estimate on when this PR will be merged. We are really looking forward to this longed-for feature :-)

**Comment by mathewantony31** [comment]
@ghalter, thanks for your interest! We're still thinking through the details of this design and when it will become generally available. We're aiming to have more details on our plan by the summer (late June).

**Comment by shermify** [comment]
I love this feature and am probably using it a lot more than you intended at this point lol, but I just want to give some feedback from a future end-user.  It works perfectly for applying color maps to grayscale imagery.  Like you said there are some issues with categorical data, but generally resolved with 'nearest' resampling.  However, that is a bit buggy on zooming in and out, etc.  I would say categorical data is going to be a very common use case so probably worth getting right.  Super stoked for this feature, not sure how much I can contribute, but I have a bunch of complex use cases up and running if you need help testing.

**Comment by rreusser** [comment]
Thanks for the feedback, @shermify! It's very helpful! 🙇 In general I think there's a lot to be said for the "less is more" philosophy, that properties should be added sparingly and not every conceivable configuration option needs to be configurable. The categorical use case does keep arising though, and I think it mostly comes down to the distinction between colorize then interpolate (categorical) vs. interpolate then colorize (quantitative). I wonder if something like a `"raster-color-mode": "quantitative" | "categorical"` style spec property would be a sensible and necessary way to distinguish between the two.

**Comment by shermify** [comment]
I agree, simpler is better.  I don't know if it even needs to be explicitly supported as a mode if it can work with tweaking existing settings.  I have a map with lots of categorical data that needs to be mapped to many colors and you can see how the issue crops up.

Screenshot 2023-04-19 at 6 29 53 PM

However, setting raster-resampling to "nearest" works a bit better.  You also need to set raster-fade-duration to 0 or it will "flash" the boundaries when tiles load in and it seems like there are some small rendering bugs when zooming in and out where the boundaries will suddenly pop in and "stick" for some reason.  Other than that it seems to work fine.  Performance is great even with a lot of categorical data and color.

Screenshot 2023-04-19 at 6 26 45 PM

**Comment by michaelbayday** [comment]
hey folks, us at cruise are greatly interested in this functionality and would appreciate a new version of mapbox to include this

**Comment by shermify** [comment]
FYI for anyone following this issue, this feature is currently in Mapbox v3.0.0-beta.  It was merged in via a different branch.

https://github.com/mapbox/mapbox-gl-js/commit/817fe631d3228c3e732f8f5a8411b66437158891

**Comment by rreusser** [comment]
Thanks for pointing this out, @shermify! I'm indeed going to close this PR as the work made its way into the beta through a different commit. 🎉

**Pull request closed without merging**
